lib/senders/index.js

```diff
@@ -24,143 +21,228 @@ module.exports = function (config, log) {
-            uaBrowserVersion: opts.uaBrowserVersion,
-            uaOS: opts.uaOS,
-            uaOSVersion: opts.uaOSVersion
-          }
-        ))
-      }
-      mailer.sendPasswordChangedNotification = function (email, opts) {
-        return P.resolve(mailer.passwordChangedEmail(
-          {
-            email: email,
-            acceptLanguage: opts.acceptLanguage || defaultLanguage,
-            ip: opts.ip,
-            location: opts.location,
-            uaBrowser: opts.uaBrowser,
-            uaBrowserVersion: opts.uaBrowserVersion,
-            uaOS: opts.uaOS,
-            uaOSVersion: opts.uaOSVersion
-          }
-        ))
-      }
-      mailer.sendPasswordResetNotification = function (email, opts) {
-        return P.resolve(mailer.passwordResetEmail(
-          {
-            email: email,
-            acceptLanguage: opts.acceptLanguage || defaultLanguage,
-            flowId: opts.flowId,
-            flowBeginTime: opts.flowBeginTime,
-          }
-        ))
-      }
-      mailer.sendNewDeviceLoginNotification = function (email, opts) {
-        return P.resolve(mailer.newDeviceLoginEmail(
-          {
-            acceptLanguage: opts.acceptLanguage || defaultLanguage,
-            flowId: opts.flowId,
-            flowBeginTime: opts.flowBeginTime,
-            email: email,
-            ip: opts.ip,
-            location: opts.location,
-            timeZone: opts.timeZone,
-            uaBrowser: opts.uaBrowser,
-            uaBrowserVersion: opts.uaBrowserVersion,
-            uaOS: opts.uaOS,
-            uaOSVersion: opts.uaOSVersion
-          }
-        ))
-      }
-      mailer.sendPostVerifyEmail = function (email, opts) {
-        return P.resolve(mailer.postVerifyEmail(
-          {
-            email: email,
-            acceptLanguage: opts.acceptLanguage || defaultLanguage
+  .then(function (senders) {
+    var ungatedMailer = senders.email
+    var configBounces = config.bounces || {}
+    var MAX_HARD = configBounces.hard && configBounces.hard.max || 0
+    var MAX_SOFT = configBounces.soft && configBounces.soft.max || 0
+    var MAX_COMPLAINT = configBounces.complaint && configBounces.complaint.max || 0
+    var DURATION_HARD = configBounces.hard && configBounces.hard.duration || Infinity
+    var DURATION_SOFT = configBounces.soft && configBounces.soft.duration || Infinity
+    var DURATION_COMPLAINT = configBounces.complaint && configBounces.complaint.duration || Infinity
+    var BOUNCE_TYPE_HARD = 1
+    var BOUNCE_TYPE_SOFT = 2
+    var BOUNCE_TYPE_COMPLAINT = 3
+
+    function checkBounce(bounce, rules, now) {
+      for (var i = 0; i < rules.length; i++) {
+        var ruleSet = rules[i]
+
+        if (bounce.bounceType === ruleSet.type) {
+          if (bounce.createdAt > now - ruleSet.duration) {
+            ruleSet.count++
+            if (ruleSet.count > ruleSet.max) {
+              throw ruleSet.error()
+            }
           }
-        ))
+          break
```

Comment:
You could avoid this `for...break` idom by indexing `rules` by the bounce type, then doing a direct lookup of `rules[bounce.bounceType].count`.